cmd/bbs/db_health_check_runner.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"code.cloudfoundry.org/bbs/db"
+	"code.cloudfoundry.org/clock"
+	"code.cloudfoundry.org/lager/v3"
+)
+
+type DBHealthCheckRunner struct {
+	logger                      lager.Logger
+	sqlDB                       db.BBSHealthCheckDB
+	clock                       clock.Clock
+	migrationsDone              chan struct{}
+	HealthCheckFailureThreshold int
+	HealthCheckTimeout          time.Duration
+	HealthCheckInterval         time.Duration
+}
+
+func NewDBHealthCheckRunner(logger lager.Logger, sqlDB db.BBSHealthCheckDB, clock clock.Clock, failureCount int, timeout, interval time.Duration, migrationsDone chan struct{}) *DBHealthCheckRunner {
+	if failureCount == 0 {
+		failureCount = 3
+	}
+	if timeout == 0 {
+		timeout = 5 * time.Second
+	}
+	if interval == 0 {
+		interval = 10 * time.Second
+	}
+	return &DBHealthCheckRunner{
+		logger:                      logger.Session("db-health-check-runner"),
+		sqlDB:                       sqlDB,
+		clock:                       clock,
+		HealthCheckFailureThreshold: failureCount,
+		HealthCheckTimeout:          timeout,
+		HealthCheckInterval:         interval,
+		migrationsDone:              migrationsDone,
+	}
+}
+
+func (runner *DBHealthCheckRunner) Run(signals <-chan os.Signal, ready chan<- struct{}) error {
+	close(ready)
+	runner.logger.Info("waiting-for-db-migrations")
+	select {
+	case <-runner.migrationsDone:
+	case <-signals:
+		runner.logger.Debug("exiting-due-to-signal")
+		return nil
+	}
+	runner.logger.Info("starting")
+	defer runner.logger.Info("exiting")
+	ticker := runner.clock.NewTicker(runner.HealthCheckInterval)
+	for {
+		runner.logger.Debug("reentering-run-loop")
+		select {
+		case <-ticker.C():
+			runner.logger.Debug("executing-health-check")
+			err := runner.ExecuteTimedHealthCheckWithRetries()
+			if err != nil {
+				runner.logger.Error("catastrophic-database-failure-detected", err)
+				return err
+			}
+			runner.logger.Debug("health-check-succeeded")
+
+		case <-signals:
+			runner.logger.Debug("exiting-due-to-signal")
+			return nil
+		}
+	}
+}
+
+func (runner *DBHealthCheckRunner) ExecuteTimedHealthCheckWithRetries() error {
+	var errs []error
+	var i int
+	for i = range runner.HealthCheckFailureThreshold {
```

Comment:
Range 5 will results in the numbers 0...4. 

When talking about attempts I think you should add 1 like we do in other places, so that the first attempt is attempt "1".

Reply:
Can you add a test that includes the attempt # log? I don't think it is fixed by looking at this.